### ISSUE TITLE
Integrate ReasonBERT 

### ISSUE THREAD

**Issue by bhavani105** [primeqa/primeqa#304]
**ReasonBERT**
Paper: https://aclanthology.org/2021.emnlp-main.494.pdf
Github: https://github.com/sunlab-osu/ReasonBERT
Huggingface: https://huggingface.co/osunlp

Datasets evaluated the model on in this paper: 6 datasets in MRQA, HotpotQA, NQTables (Natural Questions that can be answered by a table), HybridQA. 

Adding ReasonBERT to PrimeQA in order to enable researchers to conveniently compare ReasonBERT  with other baselines on these or more QA datasets and continue to train ReasonBert on their own datasets for their own purposes. 

We could also incorporate our pre-training corpus and pre-training code which people can easily adapt to pre-train their own model, if that is within the PrimeQA scope.

**Comment by xiang-deng** [comment]
The checkpoints `osunlp/ReasonBERT-RoBERTa-base` and `osunlp/ReasonBERT-TAPAS-base` are consistent with the original RoBERTa and TAPAS models. I have tested the model with the default QA header in Transformers, the performance drops under the extreme Few-shot setting. e.g., with 16 examples, but perform on par with the original implementation used in the paper (We use a matching style QA header compared to the classification style in Transformers). Will need to further test its performance with PrimeQA codebase.

**Comment by bhavani105** [comment]
@xiang-deng Thanks very much for the update.  Could you please post the script/arguments used in your experiments and the results in this issue.  I would also like to start running some experiments.

**Comment by avisil** [comment]
Can you please add what hyper-parameters were used @xiang-deng ? e.g. number of epochs for fine-tuning, learning rate, which seeds were tried. Thanks.

**Comment by bhavani105** [comment]
Here are results on 5 MRQA datasets running PrimeQA with ReasonBERT compared to result in the paper.
The experiments were run with ALL examples and 128 randomly selected examples

**All the PrimeQA score are lower.**  

LR: 4e-5
Num Epochs: 2 when using ALL training examples, best of 100 or 200 epochs using 128 examples
MRQA Preprocessor in branch add_mrqa https://github.com/primeqa/primeqa/tree/add_mrqa


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/bsiyer/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/bsiyer/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Model | Train Size | SQuAD | TriviaQA | NQ | NewsQA | SearchQA
-- | -- | -- | -- | -- | -- | --
RoBERTa (paper) | all | 92 | 78.1 | 80.6 | 71.9 | 85.2
ReasonBERT (paper) | all | 91.4 | 78.9 | 80.8 | 71.4 | 85.3
RoBERTa (paper) | 128 | 48.8 +-4.2 | 36.0 +-2.9 | 36.4 +-2 | 22.8 +02.4 | 41.3 +-2
ReasonBERT (paper) | 128 | 66.7 +-2.9 | 62.1 +-0.9 | 49.8 +-1.6 | 35.7 +-1.5 | 62.3 +-1.7
RoBERTa (primeqa) | all | 90.0 | 74.4 | 79.1 | 71.7 | 79.6
ReasonBERT (primeqa) | all | 88.8 | 74.5 | 78.5 | 69.3 | 79.0
RoBERTa (primeqa) | 128 | 41.2 | 21.1 | 34.5 | 18.1 | 29.8
ReasonBERT (primeqa) | 128 | 59.4 | 46.7 | 46.7 | 32.5 | 38.2



</body>

</html>


Example Command using MRQA Preprocessor on SQUAD:
`python /dccstor/bsiyer6/public/primeqa/primeqa/mrc/run_mrc.py --model_name_or_path osunlp/ReasonBERT-RoBERTa-base --dataset_name mrqa --dataset_config_name plain_text 
--dataset_filter_column_value SQuAD 
--dataset_filter_column_name subset 
--max_train_samples 128 
--preprocessor primeqa.mrc.processors.preprocessors.mrqa.MRQAPreprocessor 
--postprocessor primeqa.mrc.processors.postprocessors.squad.SQUADPostProcessor 
--eval_metrics SQUAD 
--output_dir /dccstor/bsiyer6/public/experiments/MRQA_SQuAD/train-reasonbert-base-128_20ep -
-overwrite_output_dir --do_train --do_eval --fp16 --learning_rate 4e-5 --per_device_train_batch_size 32 
--per_device_eval_batch_size 128 --gradient_accumulation_steps 2 --warmup_ratio 0.1 --weight_decay 0.1 --save_steps 50000 --cache_dir /dccstor/bsiyer6/.cache/huggingface --overwrite_cache --num_train_epochs 20 --evaluation_strategy no --single_context_multiple_passages`

**Comment by xiang-deng** [comment]
Here is the results I got with original ReasonBERT code + Huggingface default QA header
The hyperparameters are set the same as in the paper, **LR: 5e-5, with warm-up and decay**, the results are average over 4 runs with different random seeds.
|Model|Train Size|SQuAD|TriviaQA|SearchQA|
|-|-|-|-|-|
|RoBERTa|128|53.1+-3.9|35.6+-0.8|36.5+-3.3|
|ReasonBERT|128|66.9+-2.4|62.1+-1.7|60.6+-1.5|

Another implementation difference that I can think of is that we use the [eval code](https://github.com/mrqa/MRQA-Shared-Task-2019/blob/master/mrqa_official_eval.py) from MRQA shared task where they have some normalization over the answers.

**Comment by bhavani105** [comment]
> 

With HF header, SearchQA is close to 5 points lower for RoBERTa and 2 points for ReasonBERT - do you have any idea why that is

**Comment by bhavani105** [comment]
Here are updated PrimeQA results:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/bsiyer/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/bsiyer/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Model | Train Size | SQuAD | TriviaQA | NQ | NewsQA | SearchQA
-- | -- | -- | -- | -- | -- | --
RoBERTa (paper) | all | 92 | 78.1 | 80.6 | 71.9 | 85.2
ReasonBERT (paper) | all | 91.4 | 78.9 | 80.8 | 71.4 | 85.3
RoBERTa (paper) | 128 | 48.8 +-4.2 | 36.0 +-2.9 | 36.4 +-2 | 22.8 +02.4 | 41.3 +-2
ReasonBERT (paper) | 128 | 66.7 +-2.9 | 62.1 +-0.9 | 49.8 +-1.6 | 35.7 +-1.5 | 62.3 +-1.7
RoBERTa (paper) | 1024 | 77.9+-0.5 | 62.2+-1.3 | 60.3+-0.5 | 55+-0.3 | 67.5+-0.8
ReasonBERT (paper) | 1024 | 79.7+-0.3 | 70.1+-0.2 | 65+-0.9 | 54.7+-0.6 | 72.8+-0.8
RoBERTa (primeqa) | all | 90.0 | 74.4 | 79.1 | 71.7 | 79.6
ReasonBERT (primeqa) | all | 88.8 | 74.5 | 78.5 | 69.3 | 79.0
RoBERTa (primeqa) | 128 | 50.9+-2.1 | 21+-2.2 | 33.9+-1.5 | 23.2+-3.5 | 28.6+-1.5
ReasonBERT.(primeqa) | 128 | 64.3+-0.8 | 50.4+-1.2 | 47.3+-0.3 | 36.6+-2.0 | 49.5+-2.5
RoBERTa (primeqa) | 1024 | 76.6+-0.4 | 49+-1.5 | 59.8+-0.9 | 54.4+-0.9 | 55.7+-1.7
ReasonBERT (primeqa) | 1024 | 77.1+-0.4 | 60+-0.9 | 64+-0.2 | 52.4+-0.5 | 64.2+-0.7



</body>

</html>

Parameters in few-shot setting:
`--preprocessor primeqa.mrc.processors.preprocessors.mrqa.MRQAPreprocessor  
 --postprocessor primeqa.mrc.processors.postprocessors.squad.SQUADPostProcessor 
 --per_device_train_batch_size 20  
 --negative_sampling_prob_when_has_answer 1.0 
 --negative_sampling_prob_when_no_answer 0.1
 --learning_rate 5e-5
 --num_train_epochs 20`.

Additionally for SQuAD
`--max_seq_length 384`

**Comment by xiang-deng** [comment]
Some other things to try may be that, we use a max_seq_length of 512 (include question), and stride of 128 for long passages. Also I kept all chunks of input, i.e. without negative sampling. Though I have tested different negative sampling ratios and didn't notice a huge difference in end accuracy.

**Comment by bhavani105** [comment]
> Some other things to try may be that, we use a max_seq_length of 512 (include question), and stride of 128 for long passages. Also I kept all chunks of input, i.e. without negative sampling. Though I have tested different negative sampling ratios and didn't notice a huge difference in end accuracy.

Thanks @xiang-deng I have used max_seq_length of 512 (include question) except for SQuAD set to 384, and stride of 128 and used all input without negative sampling.  
Is there anything else that you are doing to construct the features.

**Comment by xiang-deng** [comment]
The reader is implemented [here](https://github.com/sunlab-osu/ReasonBERT/blob/eb8f48f5fc1cc7d5733edf024abb25e58681c87a/data_loader/data_loaders.py#L808), it is just standard tokenization and chunking.

Given all datasets except for TriviaQA and SearchQA have almost the same performance as we reported in the paper, despite the use of different headers, I guess it is caused by some implementation related to the reader & answer aggregation. From what was reported in the original MRQA paper, these two datasets are the long ones (>700 tokens) compared to the rest (150 ~ 600) thus the context needs to be chunked and later aggregate the scores from different chunks.

Meanwhile, another thing to note by 128/1024 training examples, I mean the original examples before processing, for long paragraphs, the chunking will lead to slightly more training examples.
Some other details

- The aggregation of results is [here](https://github.com/sunlab-osu/ReasonBERT/blob/eb8f48f5fc1cc7d5733edf024abb25e58681c87a/trainer/trainer.py#L156) and [here](https://github.com/sunlab-osu/ReasonBERT/blob/eb8f48f5fc1cc7d5733edf024abb25e58681c87a/trainer/trainer.py#L222). 
- We use the [CLS] token in the beginning to find unanswerable cases, so the score is given as ` start_pred_logits = start_pred_logits - start_pred_logits[:, :, 0:1]`
- I notice that the tokenizer of transformers does not always `decode` to the original text (some whitespaces along with others, may have been fixed). So I will also tokenize and then decode the original answer as ground truth
